### PR TITLE
implement MIDI features: pitch bend, (N)RPN tracking, bend sensitivity, hold pedal

### PR DESCRIPTION
**Pull request by jpcima**
#19 

I implement pitch bend functionality.
On a bend event, I update the bend value, go over the active channels and I invoke the same code as `PlayNoteF` as previously to update frequency registers.  
Except now I split that relevant part into a function `PlayNoteCh` which applies to an existing channel number.

No support for RPN in this commit

### CODE REVIEW

**Comment by Wohlstand** [comment]
PRN needs to capture two events and then apply it's value in final.
Also, please reset the bend into initial state when, for example, changing the instrument OR toggling the MIDI input. Alternatively, why not to give the UI to change & display current pitch bend value from the GUI even without of MIDI input?

**Comment by jpcima** [comment]
It's not about not knowing RPN, just me advancing slow and working in small incremental commits.
I was busy on few other things.

To reset the bend yeah I agree. I could get to work also on UI stuff regarding this piano thing and related. The thing is: I never use it because for me, it doesn't work right. In this program I have an auto-repeat issue not detected in Qt, making the thing unusable.

Other than that, the piano has other annoyances. A mouse press doesn't allow to move across keys while clicked. Also it doesn't have fixed horizontal aspect ratio, so it looks all squished in the import dialog.

You're aware of the VMPK software? It has a piano widget made in Qt which is reusable.
I can use this on my computer and it doesn'y suffer auto-repeat.
(widget is called vpiano I think)

**Comment by jpcima** [comment]
The UI control is added. It's a vertical slider on the keyboard's left side.
I've got to test bend sensitivity. For that, my keyboard doesn't support sending this RPN, and I must find software which does this because I don't know one right now.

**Comment by beastlike** [comment]
There's a way to reroute cc messages from other knobs/sliders on your midi controller to pitch bend/mod wheel messages in Reaper. Resort also supports joystick/gamepad analog knobs to be programmed to be a pitch bend/mod wheel. Works ok for pitch bend because they reset to zero.

Worst case scenario you can use another machine with Reaper, a game pad and a midi out, with a midi cable connected to your test machine's midi in port; although there may be a way to do it all on the same machine. 

Will test on my keyboard as soon as I get the chance

**Comment by jpcima** [comment]
Bend sensitivity works. Tested with: https://github.com/gbevin/SendMIDI

**Comment by jpcima** [comment]
I implemented hold pedal according to my understanding of MIDI spec.
I tested by emulating pedal on mod.wheel because I don't actually have one.
Right now, there won't probably be a lot of use in it, because all the notes are shut down after editing a parameter.

**Comment by Wohlstand** [comment]
I'll pull this and review at my side

File: src/opl/generator.h [hunk 1]
```diff
@@ -142,12 +151,22 @@ class Generator
...
-        uint8_t noteOn(int note);
+        uint8_t noteOn(int note, bool *replace = nullptr);
         int8_t  noteOff(int note);
+        void    channelOff(int ch);
+        int8_t  findNoteOffChannel(int note);
+        void hold(int ch, bool h);
         void clearNotes();
+        const Note &channel(int ch) const
+            { return channels.at(ch); }
+        const int channelCount() const
```
Comment (Wohlstand) [inline src/opl/generator.h]: GCC 6 says me `const` is junk here.
So, you are free to define
```
int channelCount() const
```
as returning type is not a pointer or even not a reference :wink: 
I have fixed it on my side, and if everything is fine, I'll push my state to master

Comment (jpcima) [reply src/opl/generator.h]: I have no idea why I have even written that. The fingers act before brain thinks sometimes.

**Comment by jpcima** [comment]
Good. :+1: 

> that would to go as separate task

Let's keep a list of these things to implement.
There can be a "MIDI Implementation Chart" type document to keep track. (libADLMIDI too)
I usually use this to implement synthesizer VST myself, as MIDI TODO list.

**Comment by Wohlstand** [comment]
About of the PIano widget - it's the thing purely made by me, and yeah, you are right, I must to make the support for sliding between keys while mouse is held. Piano is only a small helper to easily test lowest and highest notes, or just tick a random note by the mouse

**Comment by Wohlstand** [comment]
EDIT: Made just now: https://github.com/Wohlstand/OPL3BankEditor/commit/dd657b8ea9c6ad0101d39f3173569eadf3103f61

**Comment by jpcima** [comment]
Cool. It works! This did not take long to make.

What do you say that I make a MIDI chart document in spreadsheet format, and I put various color code for features. (todo, done, partial, requested, etc..)

**Comment by Wohlstand** [comment]
About of velocities, their I also have added just now :wink:

**Comment by jpcima** [comment]
It works :+1:

**Comment by jpcima** [comment]
The silence functionality seems now broken. Is it me or you have changed something in this?
I can't shut up the notes.

**Comment by jpcima** [comment]
Forget, it's my brain which doesn't work... Just got mixed in synths and midi connections..

**Comment by Wohlstand** [comment]
The "shut up" must clean up all "notes are on" and mute all physical chip channels. You can check the thing. However, I now gonna to sleep, therefore most of stuff I'll check tomorrow :wink:

**Comment by Wohlstand** [comment]
> Forget, it's my brain which doesn't work... Just got mixed in synths and midi connections..

Do you mean you have ran two parallel ports are stealing each other and when you shuting up one synth, second still play?

**Comment by jpcima** [comment]
I use Jack midi patchbay, and the synth midi was routed into several instruments, and the sound I was receiving in my ear was ADLMIDI but not from a different instance of the Editor. :man_facepalming: 
Of course it wouldn't respond to silence commands..

A thing though, is that the Editor could be made to respond to Program Change messages, and make UI react accordingly. It would take a second FIFO of RT->UI, with some message checking interval (like VST, where a 50ms interval is typical).

**Pull request merged**